### PR TITLE
Add functionality to top off battery without changing underlying BPM mode.

### PR DESCRIPTION
**Pull request by DimseBoms**
Added functionality to temporarily charge the battery to 100% without changing the underlying BPM mode. Quite useful if you mostly work at home but need a full charge once in a while. The old BPM mode will be reinstated if the charger is disconnected or if the battery completes a full charge.

The function is only available if the charger is connected and BPM is not disabled.

Since it reads values directly from /sys and /etc it's going to know if the extension has been restarted and it will restart the battery watcher if it notices that the Top_off function is supposed to be active.

bilde

### CODE REVIEW

File: extension.js [hunk 1]
```diff
@@ -154,17 +168,14 @@ class HuaweiWmiIndicator extends PanelMenu.Button { // TODO: move to system batt
...
 
 	_set_bpm(low, high) {
-		let file_sys = Gio.File.new_for_path("/sys/devices/platform/huawei-wmi/charge_control_thresholds");
-		let file_def = Gio.File.new_for_path("/etc/default/huawei-wmi/charge_control_thresholds");
-
 		if (low || high)
 			try {
-				file_sys.replace_contents(`${low} ${high}`, null, false, 0, null);
-				file_def.replace_contents(`${low} ${high}`, null, false, 0, null);
+				this._file_sys.replace_contents(`${low} ${high}`, null, false, 0, null);
```
Comment (egormanga) [inline extension.js]: Would this still work after the module reinserted without reopening the file?

File: extension.js [hunk 2]
```diff
@@ -174,6 +185,98 @@ class HuaweiWmiIndicator extends PanelMenu.Button { // TODO: move to system batt
 		this._bpm.label.set_text(this._BPM + `: ${low}%-${high}%`);
 	}
 
+	_update_top_off() {
+		this._get_battery(proxy => {
+			let is_discharging = proxy.State === UPower.DeviceState.DISCHARGING
+			let is_fully_charged = proxy.State === UPower.DeviceState.FULLY_CHARGED
```
Comment (egormanga) [inline extension.js]: Code style: suggest adding parentheses around the comparisons.

File: extension.js [hunk 3]
```diff
@@ -174,6 +185,98 @@ class HuaweiWmiIndicator extends PanelMenu.Button { // TODO: move to system batt
...
+			else if (is_discharging) this._stop_top_off();
+		})
+	}
+
+	_start_top_off() {
+		this._get_battery(proxy => {  // Connects watcher
+			this._battery_watching = proxy.connect(
+				'g-properties-changed',
+				this._update_top_off.bind(this)
+			);
```
Comment (egormanga) [inline extension.js]: Code style: I'd like it more non-line-wrapped.

File: extension.js [hunk 4]
```diff
@@ -174,6 +185,98 @@ class HuaweiWmiIndicator extends PanelMenu.Button { // TODO: move to system batt
...
+				this._topping_off = true;
+			} catch (e) {}
+		})
+	}
+
+	_stop_top_off() {
+		let def_low, def_high;
+		this._get_battery(proxy => {  // Disconnects watcher
+			proxy.disconnect(this._battery_watching);
+			try {	// Reinstates old BPM values
```
Comment (egormanga) [inline extension.js]: Code style: the comment here is separated with a tab instead of 2 spaces.

File: extension.js [hunk 5]
```diff
@@ -174,6 +185,98 @@ class HuaweiWmiIndicator extends PanelMenu.Button { // TODO: move to system batt
...
+				[def_low, def_high] = ByteArray.toString(this._file_def.load_contents(null)[1]).split(' ').map(Number);
+				this._file_sys.replace_contents(`${def_low} ${def_high}`, null, false, 0, null);
+				this._topping_off = false;
+			} catch (e) {}
+			})
```
Comment (egormanga) [inline extension.js]: Code style: indentation?

File: extension.js [hunk 6]
```diff
@@ -174,6 +185,98 @@ class HuaweiWmiIndicator extends PanelMenu.Button { // TODO: move to system batt
...
+		this._get_battery(proxy => { is_charging = proxy.State === UPower.DeviceState.CHARGING });
+		
+		// Handle state change
+		if (state !== undefined)
+		try {
+			if (state && !this._topping_off && is_charging) this._start_top_off();  // Top off switch gets switched on
+			else if (!state && this._topping_off) this._stop_top_off(); 		    // Top off switch gets switched off
+		} catch (e) {
+			global.log(e)
+		}
```
Comment (egormanga) [inline extension.js]: Code style: indentation; 2 spaces instead of a tab to separate the comments.

File: extension.js [hunk 7]
```diff
@@ -174,6 +185,98 @@ class HuaweiWmiIndicator extends PanelMenu.Button { // TODO: move to system batt
...
+		try {
+			[sys_low, sys_high] = ByteArray.toString(this._file_sys.load_contents(null)[1]).split(' ').map(Number);
+			[def_low, def_high] = ByteArray.toString(this._file_def.load_contents(null)[1]).split(' ').map(Number);
+
+			// If BPM == off -> unavailable
+			if (def_low == 0 && def_high == 100) {
+				this._top_off.setToggleState(false);
+				this._top_off.setSensitive(false);
+			}
+            // If BPM IS on AND device is charging -> available
```
Comment (egormanga) [inline extension.js]: Code style: indentation (the code uses tabs, here are spaces).

File: extension.js [hunk 8]
```diff
@@ -174,6 +185,98 @@ class HuaweiWmiIndicator extends PanelMenu.Button { // TODO: move to system batt
...
+				if (top_is_active && is_charging) {
+					this._top_off.setToggleState(true);
+					if (top_is_active && !this._topping_off) this._start_top_off(); // Reconnects watcher if extension has been restarted without reinstating BPM
+				} // If top is active and charger is not connected -> reinstate old BPM
+				else if (top_is_active && !is_charging) {
+					this._stop_top_off();
+					this._top_off.setToggleState(false);
+					this._top_off.setSensitive(false);
+				} 
+				} else { // Reinstates old BPM in case of unclean watcher exit and handles edge cases
```
Comment (egormanga) [inline extension.js]: Code style: 2 spaces instead of 1 to separate comments.

File: extension.js [hunk 9]
```diff
@@ -234,3 +337,4 @@ function init(meta) {
 
 // by Sdore, 2021-22
 //   apps.sdore.me
+
```
Comment (egormanga) [inline extension.js]: Code style: just why... please remove this line.

**Review by egormanga — CHANGES_REQUESTED** [review]
LGTM otherwise. Will accept after the fixes. Thank you!

File: extension.js (hunk 1 above)
Comment (DimseBoms) [reply extension.js]: I don't know what you mean about reloading the module but the files stays open as instance variables on line 43 and 44. It is done this way since there is now multiple methods where the files are going to be used.

One issue that I see now is that the If on 160/171 is not wrapped in a block..

How would I go about editing these requested changes. Do I just commit them to my fork?

Comment (egormanga) [reply extension.js]: I mean `rmmod`'ing then `modprobe`'ing the [kernel module](https://github.com/aymanbagabas/Huawei-WMI), e.g. during its development. The file descriptor will then become stale and unusable, that's the exact reason why I reopen them each time in the first place.

The `if` on L171 does not need its content wrapped, as it's a single statement (a `try..catch` block). Per my code style, the braces are redundant in this case. Please remove them, too.

And yes, you should just commit and push your changes to the same branch — GitHub will handle the rest.

File: extension.js [hunk 10]
```diff
@@ -174,6 +186,95 @@ class HuaweiWmiIndicator extends PanelMenu.Button { // TODO: move to system batt
...
+	_stop_top_off() {
+		let def_low, def_high;
+		this._get_battery(proxy => {  // Disconnects watcher
+			proxy.disconnect(this._battery_watching);
+			try {  // Reinstates old BPM values
+				[def_low, def_high] = ByteArray.toString(this._file_def.load_contents(null)[1]).split(' ').map(Number);
+				this._file_sys.replace_contents(`${def_low} ${def_high}`, null, false, 0, null);
+				this._topping_off = false;
+			} catch (e) {}
+		})
```
Comment (egormanga) [inline extension.js]: There's a missing `;` in this line (to close `this._get_battery(`).

File: extension.js [hunk 11]
```diff
@@ -233,4 +334,4 @@ function init(meta) {
 }
 
 // by Sdore, 2021-22
-//   apps.sdore.me
+//   apps.sdore.me
```
Comment (egormanga) [inline extension.js]: Now, there should be a newline — just a single, not two as you did earlier. :)

**Review by egormanga — CHANGES_REQUESTED** [review]

File: extension.js (hunk 1 above)
Comment (DimseBoms) [reply extension.js]: If that is the case then I'm going to rework the code to use the previous method of loading files. I did not even realise that this could be an issue as I'm still a novice. Thank you for pointing this out :)

Comment (egormanga) [reply extension.js]: Okay, that's kind of correct now. I'll probably rework variable naming later; good for now.

**Review by egormanga — APPROVED** [review]

**Comment by egormanga** [comment]
@sermart1234 Please test that PR on your device model.

File: extension.js [hunk 12]
```diff
@@ -283,6 +270,16 @@ class HuaweiWmiIndicator extends PanelMenu.Button { // TODO: move to system batt
...
+
+		// Handle state change
+		if (state !== undefined)
+		try {
+			if (state && !this._topping_off) this._start_top_off();  // Top off switch gets switched on
+			else if (!state && this._topping_off) this._stop_top_off();  // Top off switch gets switched off
+			this._set_top_off();
+		} catch (e) {
+			global.log(e)
+		}
```
Comment (egormanga) [inline extension.js]: Please indent that block properly (+1 tab).

File: extension.js [hunk 13]
```diff
@@ -236,43 +236,30 @@ class HuaweiWmiIndicator extends PanelMenu.Button { // TODO: move to system batt
...
 				this._top_off.setSensitive(true);
-				// Check if top-off is active and set button state
+				// Check if top-off is active -> Button = On
 				let top_is_active = ((def_low != 0 && def_high != 100) && (sys_low == 0 && sys_high == 100));
-				if (top_is_active && is_charging) {
+				if (top_is_active) {
 					this._top_off.setToggleState(true);
 					if (top_is_active && !this._topping_off) this._start_top_off();  // Reconnects watcher if extension has been restarted without reinstating BPM
-				}  // If top is active and charger is not connected -> reinstate old BPM
-				else if (top_is_active && !is_charging) {
```
Comment (egormanga) [inline extension.js]: Please join these two lines, as per my code style.

File: extension.js [hunk 14]
```diff
@@ -236,43 +236,30 @@ class HuaweiWmiIndicator extends PanelMenu.Button { // TODO: move to system batt
...
-					this._stop_top_off();
-					this._top_off.setToggleState(false);
-					this._top_off.setSensitive(false);
-				}
+				}  // If top is not active -> Button = Off
+				else this._top_off.setToggleState(false);
```
Comment (egormanga) [inline extension.js]: Please join these two lines, as per my code style.

**Review by egormanga — CHANGES_REQUESTED** [review]
Perfectly fine otherwise, will merge.

**Comment by egormanga** [comment]
Sorry for the delay. @sermart1234 has become unavailable lately, so I'll have to test this myself, thus on my laptop model only.

File: extension.js [hunk 15]
```diff
@@ -174,6 +188,99 @@ class HuaweiWmiIndicator extends PanelMenu.Button { // TODO: move to system batt
...
+				let top_is_active = ((def_low != 0 && def_high != 100) && (sys_low == 0 && sys_high == 100));
+				if (top_is_active) {
+					this._top_off.setToggleState(true);
+					if (top_is_active && !this._topping_off) this._start_top_off();  // Reconnects watcher if extension has been restarted without reinstating BPM
+				} else this._top_off.setToggleState(false);  // If top is not active -> Button = Off
+				} else {  // Reinstates old BPM in case of unclean watcher exit and handles edge cases
+					if (def_low != sys_low || def_high != sys_high) this._stop_top_off();
+					this._top_off.setToggleState(false);
+					this._top_off.setSensitive(false);
+				}
```
Comment (egormanga) [inline extension.js:266]: This bit of code is not properly intended now. Should be 1 tab less.

Upd: or maybe I didn't notice it earlier.

File: extension.js [hunk 16]
```diff
@@ -174,6 +188,99 @@ class HuaweiWmiIndicator extends PanelMenu.Button { // TODO: move to system batt
...
+			[sys_low, sys_high] = ByteArray.toString(_file_sys.load_contents(null)[1]).split(' ').map(Number);
+			[def_low, def_high] = ByteArray.toString(_file_def.load_contents(null)[1]).split(' ').map(Number);
+
+			// If BPM == off -> Button = Unavailable and Off
+			if (def_low == 0 && def_high == 100) {
+				this._top_off.setToggleState(false);
+				this._top_off.setSensitive(false);
+			}
+			// If BPM is on and device is not discharging -> Button = Available
+			else if ((def_low != 0 || def_high != 100) && !is_discharging ) {
```
Comment (egormanga) [inline extension.js:254]: Remove the unnecessary space before the closing bracket at line 254 and merge these to a single line too.

In general, don't place an `else` clause with a line break after the previous `}` it corresponds to, cause that's hard to follow.

**Review by egormanga — CHANGES_REQUESTED** [review]

**Comment by egormanga** [comment]
Actually, I should've merged this immediately and then fix the style manually. But now it's too late, so fine.

**Comment by egormanga** [comment]
Yeah, I somehow forgot about this completely. Sorry for that.
